samples/gen-go-db/server/db/dynamodb/deployment.go

```diff
@@ -0,0 +1,568 @@
+	key, err := dynamodbattribute.MarshalMap(ddbDeploymentPrimaryKey{
+		EnvApp:  fmt.Sprintf("%s--%s", environment, application),
+		Version: version,
+	})
+	if err != nil {
+		return nil, err
+	}
+	res, err := t.DynamoDBAPI.GetItemWithContext(ctx, &dynamodb.GetItemInput{
+		Key:       key,
+		TableName: aws.String(t.name()),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if len(res.Item) == 0 {
+		return nil, db.ErrDeploymentNotFound{
+			Environment: environment,
+			Application: application,
+			Version:     version,
+		}
+	}
+
+	var m models.Deployment
+	if err := decodeDeployment(res.Item, &m); err != nil {
+		return nil, err
+	}
+
+	return &m, nil
+}
+
+func (t DeploymentTable) getDeploymentsByEnvAppAndVersion(ctx context.Context, input db.GetDeploymentsByEnvAppAndVersionInput, fn func(m *models.Deployment, lastDeployment bool) bool) error {
+	if input.VersionStartingAt != nil && input.StartingAfter != nil {
+		return fmt.Errorf("Must specify one of input.VersionStartingAt or input.StartingAfter")
+	}
+	queryInput := &dynamodb.QueryInput{
+		TableName: aws.String(t.name()),
+		ExpressionAttributeNames: map[string]*string{
+			"#ENVAPP": aws.String("envApp"),
+		},
+		ExpressionAttributeValues: map[string]*dynamodb.AttributeValue{
+			":envApp": &dynamodb.AttributeValue{
+				S: aws.String(fmt.Sprintf("%s--%s", input.Environment, input.Application)),
+			},
+		},
+		ScanIndexForward: aws.Bool(!input.Descending),
+		ConsistentRead:   aws.Bool(!input.DisableConsistentRead),
+	}
+	if input.Limit != nil {
+		queryInput.Limit = input.Limit
+	}
+	if input.VersionStartingAt == nil {
+		queryInput.KeyConditionExpression = aws.String("#ENVAPP = :envApp")
+	} else {
+		queryInput.ExpressionAttributeNames["#VERSION"] = aws.String("version")
+		queryInput.ExpressionAttributeValues[":version"] = &dynamodb.AttributeValue{
+			S: aws.String(*input.VersionStartingAt),
+		}
+		if input.Descending {
+			queryInput.KeyConditionExpression = aws.String("#ENVAPP = :envApp AND #VERSION <= :version")
+		} else {
+			queryInput.KeyConditionExpression = aws.String("#ENVAPP = :envApp AND #VERSION >= :version")
+		}
+	}
+	if input.StartingAfter != nil {
+		queryInput.ExpressionAttributeNames["#VERSION"] = aws.String("version")
+		queryInput.ExpressionAttributeValues[":version"] = &dynamodb.AttributeValue{
+			S: aws.String(input.StartingAfter.Version),
+		}
+		queryInput.ExclusiveStartKey = map[string]*dynamodb.AttributeValue{
+			"version": &dynamodb.AttributeValue{
+				S: aws.String(input.StartingAfter.Version),
+			},
+			"envApp": &dynamodb.AttributeValue{
+				S: aws.String(fmt.Sprintf("%s--%s", input.StartingAfter.Environment, input.StartingAfter.Application)),
+			},
+		}
+		if input.Descending {
+			queryInput.KeyConditionExpression = aws.String("#ENVAPP = :envApp AND #VERSION <= :version")
+		} else {
+			queryInput.KeyConditionExpression = aws.String("#ENVAPP = :envApp AND #VERSION >= :version")
+		}
```

Comment:
just checking--are these `KeyConditionExpression`s necessary when specifying `ExclusiveStartKey`?

If they are necessary, I think we can simply do `queryInput.KeyConditionExpression = aws.String("#ENVAPP = :envApp")`, since the `<=` or `>=` constraint is less binding than the `ExclusiveStartKey`

Reply:
nice catch!